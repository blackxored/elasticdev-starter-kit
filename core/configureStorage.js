// @flow
import { pick } from 'ramda';
import { createTransform } from 'redux-persist';

const defaultPaths = [
  ['auth'],
  ['apollo', ['data']],
  ['device'],
  ['settings'],
  ['viewer'],
];

const makeTransformsAndWhiteLists = (
  userDefinedPaths: Array<Array<string>>,
) => {
  const paths = [...defaultPaths, ...userDefinedPaths];
  const whitelist = [];
  const transforms = [];

  paths.forEach(([feature, props]) => {
    whitelist.push(feature);

    if (!props) {
      return;
    }

    const inOut = state => pick(props, state);
    transforms.push(createTransform(inOut, inOut, { whitelist: [feature] }));
  });

  return { whitelist, transforms };
};

const configureStorage = (
  appName: string,
  paths: Array<Array<string>> = [],
) => ({
  debounce: 100,
  keyPrefix: `${appName}:`,
  ...makeTransformsAndWhiteLists(paths),
});

export default configureStorage;
