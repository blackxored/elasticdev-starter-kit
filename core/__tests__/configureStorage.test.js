import configureStorage from '../configureStorage';

describe('configureStorage', () => {
  it('setups redux-persist', () => {
    expect(
      configureStorage('testapp', [['additional_store_path_to_serialize']]),
    ).toMatchSnapshot();
  });
});
