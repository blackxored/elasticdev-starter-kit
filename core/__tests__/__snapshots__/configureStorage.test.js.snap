// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configureStorage setups redux-persist 1`] = `
Object {
  "debounce": 100,
  "keyPrefix": "testapp:",
  "transforms": Array [
    Object {
      "in": [Function],
      "out": [Function],
    },
  ],
  "whitelist": Array [
    "auth",
    "apollo",
    "device",
    "settings",
    "viewer",
    "additional_store_path_to_serialize",
  ],
}
`;
